mfkdf2/src/derive/factors/hmacsha1.rs

```diff
@@ -1,36 +1,212 @@
-use std::rc::Rc;
-
-use rand::{Rng, rngs::OsRng};
+use rand::{RngCore, rngs::OsRng};
 use serde_json::{Value, json};
 
 use crate::{
-  derive::{DeriveFactorFn, factors::MFKDF2DerivedFactor},
+  crypto::{decrypt, encrypt},
+  derive::FactorDerive,
   error::MFKDF2Result,
+  setup::factors::{
+    Factor, FactorType, MFKDF2Factor,
+    hmacsha1::{HmacSha1, HmacSha1Response},
+  },
 };
 
-pub struct HMACSHA1Options {
-  pub id:     Option<String>,
-  pub secret: Option<[u8; 20]>,
-}
+impl FactorDerive for HmacSha1 {
+  fn include_params(&mut self, params: Value) -> MFKDF2Result<()> {
+    self.params = Some(serde_json::to_string(&params).unwrap());
+
+    let response = self.response.as_ref().unwrap();
+    let mut padded_key = [0u8; 32];
+    padded_key[..response.0.len()].copy_from_slice(&response.0);
+
+    let pad = hex::decode(
+      params
+        .get("pad")
+        .ok_or_else(|| crate::error::MFKDF2Error::MissingDeriveParams("pad".to_string()))?
+        .as_str()
+        .ok_or(crate::error::MFKDF2Error::InvalidDeriveParams("pad".to_string()))?,
+    )
+    .map_err(|e| crate::error::MFKDF2Error::InvalidDeriveParams(e.to_string()))?;
+
+    let padded_secret = decrypt(pad.clone(), &padded_key);
+    self.padded_secret = padded_secret;
+
+    Ok(())
+  }
 
-pub fn hmacsha1(response: [u8; 20]) -> MFKDF2Result<DeriveFactorFn> {
-  Ok(Rc::new(move |params: Value| {
-    let pad: Vec<u8> =
-      params["pad"].as_array().unwrap().iter().map(|v| v.as_u64().unwrap() as u8).collect();
-    let secret: [u8; 20] = std::array::from_fn(|i| response[i] ^ pad[i]);
-
-    Box::pin(async move {
-      Ok(MFKDF2DerivedFactor {
-        kind:   "hmacsha1".to_string(),
-        data:   secret.to_vec(),
-        params: Some(Box::new(move |_| {
-          let challenge = OsRng.r#gen::<u64>();
-          let response = crate::crypto::hmacsha1(&secret, challenge);
-          let pad = response.iter().zip(secret.iter()).map(|(a, b)| a ^ b).collect::<Vec<u8>>();
-          Box::pin(async move { json!({ "challenge": challenge, "pad": pad }) })
-        })),
-        output: Some(Box::new(move |_| Box::pin(async move { json!({ "secret": secret }) }))),
-      })
+  fn params_derive(&self, _key: [u8; 32]) -> Value {
+    let mut challenge = [0u8; 64];
+    OsRng.fill_bytes(&mut challenge);
+
+    let response = crate::crypto::hmacsha1(&self.padded_secret[..20], &challenge);
+    let mut padded_key = [0u8; 32];
+    padded_key[..response.len()].copy_from_slice(&response);
+    let pad = encrypt(&self.padded_secret, &padded_key);
+
+    json!({
+      "challenge": hex::encode(challenge),
+      "pad": hex::encode(pad),
     })
-  }))
+  }
+
+  fn output_derive(&self) -> Value {
+    json!({
+      "secret": self.padded_secret[..20],
+    })
+  }
+}
+
+impl Factor for HmacSha1 {}
```

Comment:
This is code smell imo. If this is the case, I can promise you that you basically don't need to actually have this unified implementation of a trait and it's better to have separate traits so they are invoked individually.

Reply:
removed this trait, and trait methods renamed to `params, output`.